benches/raftstore.rs

```diff
@@ -0,0 +1,202 @@
+    F: ClusterFactory<T>,
+{
+    let mut cluster = input.factory.build(input.nodes);
+    let mut kvs = generate_random_kvs(DEFAULT_DATA_SIZE, 100, 128);
+    prepare_cluster(&mut cluster, &kvs);
+
+    let mut keys = kvs
+        .drain(..)
+        .take(DEFAULT_DATA_SIZE / 10)
+        .map(|i| i.0)
+        .chain(KvGenerator::new(100, 0).map(|i| i.0));
+
+    b.iter(|| {
+        let k = keys.next().unwrap();
+        cluster.get(&k)
+    });
+}
+
+#[derive(Debug)]
+struct DeleteConfig<F> {
+    factory: F,
+    nodes: usize,
+}
+
+fn bench_delete<T, F>(b: &mut Bencher, input: &DeleteConfig<F>)
+where
+    T: Simulator,
+    F: ClusterFactory<T>,
+{
+    let mut cluster = input.factory.build(input.nodes);
+    let mut kvs = generate_random_kvs(DEFAULT_DATA_SIZE, 100, 128);
+    prepare_cluster(&mut cluster, &kvs);
+
+    let mut keys = kvs
+        .drain(..)
+        .take(DEFAULT_DATA_SIZE / 10)
+        .map(|i| i.0)
+        .chain(KvGenerator::new(100, 0).map(|i| i.0));
+
+    b.iter(|| {
+        let k = keys.next().unwrap();
+        cluster.must_delete(&k)
+    });
+}
+
+fn bench_raft_cluster<T, F>(c: &mut Criterion, factory: F)
+where
+    T: Simulator + 'static,
+    F: ClusterFactory<T>,
+{
+    let nodes_coll = vec![1, 3, 5];
+    let value_size_coll = vec![8, 128, 1024, 4096];
+
+    let mut set_inputs = vec![];
+    let mut get_inputs = vec![];
+    let mut delete_inputs = vec![];
+    for nodes in nodes_coll {
+        for &value_size in &value_size_coll {
+            set_inputs.push(SetConfig {
+                factory: factory.clone(),
+                nodes,
+                value_size,
+            });
+        }
+        get_inputs.push(GetConfig {
+            factory: factory.clone(),
+            nodes,
+        });
+        delete_inputs.push(DeleteConfig {
+            factory: factory.clone(),
+            nodes,
+        });
+    }
+    c.bench_function_over_inputs("bench_set", bench_set, set_inputs);
+    c.bench_function_over_inputs("bench_get", bench_get, get_inputs);
+    c.bench_function_over_inputs("bench_delete", bench_delete, delete_inputs);
+}
+
+trait ClusterFactory<T: Simulator>: Clone + fmt::Debug + 'static {
+    fn build(&self, nodes: usize) -> Cluster<T>;
+}
+
+#[derive(Clone)]
+struct NodeClusterFactory;
+
+impl ClusterFactory<NodeCluster> for NodeClusterFactory {
+    fn build(&self, nodes: usize) -> Cluster<NodeCluster> {
+        new_node_cluster(1, nodes)
+    }
+}
+
+impl fmt::Debug for NodeClusterFactory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Node")
+    }
+}
+
+#[derive(Clone)]
+struct ServerClusterFactory;
+
+impl ClusterFactory<ServerCluster> for ServerClusterFactory {
+    fn build(&self, nodes: usize) -> Cluster<ServerCluster> {
+        new_server_cluster(1, nodes)
+    }
+}
+
+impl fmt::Debug for ServerClusterFactory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Server")
+    }
+}
+
+fn main() {
```

Comment:
Why not `criterion_main`?

Reply:
It can be checked in `bench_raft_cluster`.

Reply:
Should we call `criterion::init_logging()` at the beginning of the benchmark?

Reply:
Haha. No! I turned it on at first. It will cause a lot of rubbish logs.

Reply:
BTW this is also another reason why criterion_main is not used.